### PR TITLE
Change layout to reference the previous sibling

### PR DESCRIPTION
**Pull request by pavpanchekha**
This changes how the layout tree is linked and the protocol for computing layout. Basically, in addition to a parent and a list of
children, each layout object now also has a pointer to its previous sibling. That way, to compute its Y-position, a layout object can reference its previous sibling's Y-position and height.

This might seem more complex than having the parent keep track of a vertical cursor for children, but the advantage of this approach is that computing the layout of each layout object is totally self-contained to a single `layout` function, instead of
responsibilities being split between that node's and its parent's `layout` functions. It also means that the recursive calls to
`layout` are purely that---calling `layout` on each child---instead of being interlaced with code that tracks a vertical cursor.

I've also added a paragraph that talks about dependencies and why the order is so important—this is easier to discuss now that it's purely an ordering problem within a single function. That could set us up to talk about dependencies better in the reflow chapter.

### CODE REVIEW

File: book/layout.md [hunk 1]
```diff
@@ -114,14 +114,15 @@ class BlockLayout:
 
 These layout objects will be the nodes of a *layout tree*. So we'll
 want layout objects to have children, a parent, and an HTML element
-that they correspond to:
+that they correspond to. Each layout object will also have a layout to
```
Comment (chrishtr) [inline book/layout.md:117]:
```suggestion
that they correspond to. Each layout object will also have a reference to
```

File: book/layout.md [hunk 2]
```diff
@@ -114,14 +114,15 @@ class BlockLayout:
 
 These layout objects will be the nodes of a *layout tree*. So we'll
 want layout objects to have children, a parent, and an HTML element
```
Comment (chrishtr) [inline book/layout.md:116]:
```suggestion
want layout objects to have references to their children, a parent, and an HTML element
```

File: book/layout.md [hunk 3]
```diff
@@ -143,18 +144,9 @@ def layout(self):
...
-
-``` {.python}
-def __init__(self, node, parent):
-    # ...
-    self.x = -1
-    self.y = -1
-    self.w = -1
-    self.h = -1
-```
+Besides the references to other layout object, each layout object also
```
Comment (chrishtr) [inline book/layout.md:147]:
```suggestion
Besides the references to other related layout objects, each layout object also
```

File: book/layout.md [hunk 4]
```diff
@@ -233,8 +220,7 @@ whether it should have block or inline contents:
 def has_block_children(self):
     for child in self.node.children:
         if isinstance(child, TextNode):
-            if not child.text.isspace():
```
Comment (chrishtr) [inline book/layout.md:236]: Why did these two lines get removed?

File: book/layout.md [hunk 5]
```diff
@@ -246,11 +232,12 @@ create child layout objects:
 ``` {.python indent=4}
 def layout(self):
     if self.has_block_children():
+        previous = None
         for child in self.node.children:
-            if isinstance(child, TextNode): continue
```
Comment (chrishtr) [inline book/layout.md:250]: Not skipping TextNode any more?

File: book/layout.md [hunk 6]
```diff
@@ -295,56 +282,87 @@ Size and position
...
 
 ``` {.python}
-self.h = y - self.y
+self.h = sum([child.h for child in self.children])
 ```
 
-That settles `BlockLayout`; let's now work on `InlineLayout`. Its `x`
-and `y` will be set by its parent, so its `layout` just needs to set
-`w` and `h`:
+Note how important the order is here. A block layout's width depends
```
Comment (chrishtr) [inline book/layout.md:334]:
```suggestion
Note how important the order is here. A block layout object's width depends
```

File: book/layout.md [hunk 7]
```diff
@@ -295,56 +282,87 @@ Size and position
...
+of its children.
 
 ``` {.python}
 class InlineLayout:
     def layout(self):
+        self.x = self.parent.x
         self.w = self.parent.w
+
+        if self.previous:
+            self.y = self.previous.y + self.previous.h
```
Comment (chrishtr) [inline book/layout.md:360]: Where did self.y get set in the old code? Don't see any line deletions in your PR for that.

File: book/layout.md [hunk 8]
```diff
@@ -354,10 +372,10 @@ document always starts in the same place it's pretty simple:
 ``` {.python}
 class DocumentLayout:
     def layout(self):
-        self.w = WIDTH
         # ...
-        child.x = self.x = 0
-        child.y = self.y = 0
+        self.w = WIDTH - 2*HSTEP
```
Comment (chrishtr) [inline book/layout.md:376]: Why the change here?

File: src/lab5.py [hunk 9]
```diff
@@ -54,143 +54,149 @@ def request(url):
...
+            parent = self.unfinished[-1] if self.unfinished else None
+            node = Element(tag, attributes, parent)
+            self.unfinished.append(node)
+
+    HEAD_TAGS = [
+        "base", "basefont", "bgsound", "noscript",
+        "link", "meta", "title", "style", "script",
+    ]
+
+    def implicit_tags(self, tag):
```
Comment (chrishtr) [inline src/lab5.py:154]: Why all these changes in parsing? Did this PR get mixed up with the previous one you wrote somehow?

File: book/layout.md (hunk 4 above)
Comment (pavpanchekha) [reply book/layout.md:236]: Thanks to there no longer being whitespace-only TextNodes, after changes in Ch4.

File: book/layout.md (hunk 5 above)
Comment (pavpanchekha) [reply book/layout.md:250]: There no won't be whitespace-only TextNodes, which is what this was skipping

File: book/layout.md (hunk 7 above)
Comment (pavpanchekha) [reply book/layout.md:360]: In the previous version, `x` and `y` were set by one's parent, and since `InlineLayout` doesn't have any children (at this stage of the book) there isn't code for `y` to delete. Honestly that was super confusing.

File: book/layout.md (hunk 8 above)
Comment (pavpanchekha) [reply book/layout.md:376]: This is unrelated to the others, actually, but without padding there seems to be some kind of border that Tk adds by default which obscures some letters, and students wondered if that was a bug in their code.

Comment (pavpanchekha) [reply book/layout.md:376]: The padding comes back next chapter anyway.

Comment (chrishtr) [reply book/layout.md:376]: OK can you add a comment explaining what's going on?

File: book/layout.md (hunk 7 above)
Comment (chrishtr) [reply book/layout.md:360]: Got it, makes sense. A nice improvement.

File: src/lab5.py (hunk 9 above)
Comment (pavpanchekha) [reply src/lab5.py:154]: I should have separated this, but yes, I had to "forward-port" the ch4 changes to the ch5 code snapshot before I made further changes.

**Review by chrishtr — APPROVED** [review]
